File: text-to-speech/voices/python/voices.py
#!/usr/bin/env python3
"""
Smallest AI Text-to-Speech - Voice Explorer

List available TTS voices, filter by language/gender/accent, and preview any voice.

Usage:
  python voices.py                             # List all voices
  python voices.py --language english          # Filter by language
  python voices.py --gender female             # Filter by gender
  python voices.py --accent american           # Filter by accent
  python voices.py --preview sophia             # Generate preview for a voice

Output:
- Table of available voices
- Optional preview WAV file
"""

import argparse
import os
import sys
import requests
from dotenv import load_dotenv

load_dotenv()

MODEL = "lightning-v3.1"
API_BASE = "https://api.smallest.ai/waves/v1"
PREVIEW_TEXT = "Hello! This is a preview of my voice. I can speak naturally with great clarity and expression."


def get_voices(api_key: str, model: str = MODEL) -> list:
    response = requests.get(
        f"{API_BASE}/{model}/get_voices",
        headers={"Authorization": f"Bearer {api_key}"},
    )

    if response.status_code != 200:
        raise Exception(f"API request failed ({response.status_code}): {response.text}")

    data = response.json()
    return data.get("voices", data) if isinstance(data, dict) else data


def filter_voices(voices: list, language: str = None, gender: str = None, accent: str = None) -> list:
    filtered = voices
    if language:
        filtered = [
            v for v in filtered
            if language.lower() in [lang.lower() for lang in v.get("tags", {}).get("language", [])]
        ]
    if gender:
        filtered = [
            v for v in filtered
            if v.get("tags", {}).get("gender", "").lower() == gender.lower()
        ]
    if accent:
        filtered = [
            v for v in filtered
            if accent.lower() in v.get("tags", {}).get("accent", "").lower()
        ]
    return filtered


def print_voices(voices: list):
    if not voices:
        print("No voices found matching your criteria.")
        return

    print(f"\n{'Voice ID':<20} {'Name':<20} {'Gender':<10} {'Languages':<25} {'Accent'}")
    print("-" * 90)

    for v in voices:
        tags = v.get("tags", {})
        voice_id = v.get("voiceId", "")
        name = v.get("displayName", "")
        gender = tags.get("gender", "—")
        languages = ", ".join(tags.get("language", []))
        accent = tags.get("accent", "—")
        print(f"{voice_id:<20} {name:<20} {gender:<10} {languages:<25} {accent}")

    print(f"\nTotal: {len(voices)} voice(s)")


def preview_voice(voice_id: str, api_key: str, model: str = MODEL):
    print(f"Generating preview for '{voice_id}'...")

    response = requests.post(
        f"{API_BASE}/{model}/get_speech",
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        json={
            "text": PREVIEW_TEXT,
            "voice_id": voice_id,
            "sample_rate": 24000,
            "speed": 1.0,
            "output_format": "wav",
        },
    )

    if response.status_code != 200:
        raise Exception(f"API request failed ({response.status_code}): {response.text}")

    filename = f"preview_{voice_id}.wav"
    with open(filename, "wb") as f:
        f.write(response.content)

    print(f"Saved to {filename} ({len(response.content):,} bytes)")


def main():
    parser = argparse.ArgumentParser(description="Explore Smallest AI TTS voices")
    parser.add_argument("--model", default=MODEL, help="TTS model (default: lightning-v3.1)")
    parser.add_argument("--language", help="Filter by language (e.g. english, hindi)")
    parser.add_argument("--gender", help="Filter by gender (male/female)")
    parser.add_argument("--accent", help="Filter by accent (e.g. american, british)")
    parser.add_argument("--preview", metavar="VOICE_ID", help="Generate a preview for a voice")
    args = parser.parse_args()

    api_key = os.environ.get("SMALLEST_API_KEY")
    if not api_key:
        print("Error: SMALLEST_API_KEY environment variable not set")
        sys.exit(1)

    if args.preview:
        preview_voice(args.preview, api_key, args.model)
        return

    voices = get_voices(api_key, args.model)
    voices = filter_voices(voices, args.language, args.gender, args.accent)
    print_voices(voices)


if __name__ == "__main__":
    main()

File: text-to-speech/voices/javascript/voices.js
#!/usr/bin/env node
/**
 * Smallest AI Text-to-Speech - Voice Explorer
 *
 * List available TTS voices, filter by language/gender/accent, and preview any voice.
 *
 * Usage:
 *   node voices.js                             # List all voices
 *   node voices.js --language english          # Filter by language
 *   node voices.js --gender female             # Filter by gender
 *   node voices.js --preview sophia             # Generate preview for a voice
 *
 * Output:
 * - Table of available voices
 * - Optional preview WAV file
 */

const fs = require("fs");

const MODEL = "lightning-v3.1";
const API_BASE = "https://api.smallest.ai/waves/v1";
const PREVIEW_TEXT =
  "Hello! This is a preview of my voice. I can speak naturally with great clarity and expression.";

function parseArgs(argv) {
  const args = { model: MODEL };
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === "--model" && argv[i + 1]) args.model = argv[++i];
    else if (argv[i] === "--language" && argv[i + 1]) args.language = argv[++i];
    else if (argv[i] === "--gender" && argv[i + 1]) args.gender = argv[++i];
    else if (argv[i] === "--accent" && argv[i + 1]) args.accent = argv[++i];
    else if (argv[i] === "--preview" && argv[i + 1]) args.preview = argv[++i];
  }
  return args;
}

async function getVoices(apiKey, model) {
  const response = await fetch(`${API_BASE}/${model}/get_voices`, {
    headers: { Authorization: `Bearer ${apiKey}` },
  });

  if (!response.ok) {
    throw new Error(
      `API request failed (${response.status}): ${await response.text()}`
    );
  }

  const data = await response.json();
  return data.voices || data;
}

function filterVoices(voices, { language, gender, accent }) {
  let filtered = voices;
  if (language) {
    filtered = filtered.filter((v) =>
      (v.tags?.language || []).some(
        (l) => l.toLowerCase() === language.toLowerCase()
      )
    );
  }
  if (gender) {
    filtered = filtered.filter(
      (v) => (v.tags?.gender || "").toLowerCase() === gender.toLowerCase()
    );
  }
  if (accent) {
    filtered = filtered.filter((v) =>
      (v.tags?.accent || "").toLowerCase().includes(accent.toLowerCase())
    );
  }
  return filtered;
}

function printVoices(voices) {
  if (!voices.length) {
    console.log("No voices found matching your criteria.");
    return;
  }

  console.log("");
  console.log(
    "Voice ID".padEnd(20) +
      "Name".padEnd(20) +
      "Gender".padEnd(10) +
      "Languages".padEnd(25) +
      "Accent"
  );
  console.log("-".repeat(90));

  for (const v of voices) {
    const tags = v.tags || {};
    console.log(
      (v.voiceId || "").padEnd(20) +
        (v.displayName || "").padEnd(20) +
        (tags.gender || "—").padEnd(10) +
        (tags.language || []).join(", ").padEnd(25) +
        (tags.accent || "—")
    );
  }

  console.log(`\nTotal: ${voices.length} voice(s)`);
}

async function previewVoice(voiceId, apiKey, model) {
  console.log(`Generating preview for '${voiceId}'...`);

  const response = await fetch(`${API_BASE}/${model}/get_speech`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      text: PREVIEW_TEXT,
      voice_id: voiceId,
      sample_rate: 24000,
      speed: 1.0,
      output_format: "wav",
    }),
  });

  if (!response.ok) {
    throw new Error(
      `API request failed (${response.status}): ${await response.text()}`
    );
  }

  const audio = Buffer.from(await response.arrayBuffer());
  const filename = `preview_${voiceId}.wav`;
  fs.writeFileSync(filename, audio);
  console.log(`Saved to ${filename} (${audio.length.toLocaleString()} bytes)`);
}

async function main() {
  const apiKey = process.env.SMALLEST_API_KEY;
  if (!apiKey) {
    console.error("Error: SMALLEST_API_KEY environment variable not set");
    process.exit(1);
  }

  const args = parseArgs(process.argv);

  if (args.preview) {
    await previewVoice(args.preview, apiKey, args.model);
    return;
  }

  const voices = await getVoices(apiKey, args.model);
  const filtered = filterVoices(voices, args);
  printVoices(filtered);
}

main();
